### ISSUE TITLE
Create rollup wide hardfork spec

### ISSUE THREAD

**Issue by eyusufatik** [chainwayxyz/citrea#235]
If anything goes wrong in production, we'd like to have a system that will allow us to fix our mistakes.

Ideally this fork system has access to state and accessory state directly (both state db and ledger db) and underlying modules can query this system like 

```rust
hard_fork_manager.is_fork1_enabled()
```

See revm and reth's hardfork managers.

More details from @eyusufatik (I guess someone added these lines besides me xd):
Create something that is accessible from many different parts of our software. All of these will probably need to know if a hardfork is activated / when it’s gonna be activated:
1. citrea evm
2. soft confirmation rule enforcer
3. STF blueprint
4. any RPC
5. anything related to storage
6. ledger db tables

reasoning:
1 and 2 are simple, change behavior of something when we need to. Set an activation block height in the future and everybody is gonna be upgraded to the version by then.
3. We might want to change the behavior of something in the sdk level.
4. obvious after all these points
5-6. when we want to create new tables etc. we’ll need to do changes without breaking anything. but these can be handled outside the “hardfork system” tho. these can be database migrations i guess.

TODOs:
- [ ] Vector of hardfork activations that map to onetime running functions, if software is run with a new hardfork immideately calls that function.
- [ ] Systemwide access to this vector is important.

**Comment by rakanalh** [comment]
# Regarding RPC:

I suggest we use versioning to make sure that our RPC calls are compatible with specific parts of the codebase.

Example:
`/v1/ledger_getHeadSoftBatch` can be different than `/v2/ledger_getHeadSoftBatch` is they use different tables to fetch data from.

Both can be maintained at the same time to support old & new clients.

That should be pretty easy to implement.

**Comment by rakanalh** [comment]
# Fork Manager

A fork manager is a component which decides when to activate a specific fork by calling a set of one-off functions that would set specific states in order for the new functionality to start running.

Assuming we define forks similar to how revem implements it:
```rust
pub enum SpecId {
   Spec1,
   Spec2,
   Spec3
}
```

At genesis, the components we have start their state at the first spec. 

But in order for us to activate a fork (spec), we need to manage how each component would migrate it's internals to work with that new speci.

```rust
/// Defines the interface of a migrations manager
pub trait ForkMigrationManager {
  /// Invoked before a spec is activated.
  fn pre_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()>;
  /// Invoked after a spec is activated.
  fn post_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()>;
}

pub NoOpMigration {};

impl ForkMigrationManager for NoOpMigration {
  fn pre_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()> {
    // Do nothing
  }

  fn post_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()> {
    // Do nothing
  }
}

```

Components like ledger DB or the EVM could implement this interface to migrate specific data based on the `spec_id` passed to the `pre` or `post` methods of this interface.

At specific blocks, a new spec will be active. So a rough fork manager will look like the following:

```rust
/// Defines the interface which a fork manager needs to implement.
pub trait ForkManagerTrait {
  /// Returns the currently active fork.
  fn active_spec(&self) -> SpecId;

  /// Activates a specific spec
  fn activate_spec(&self, spec_id: SpecId) -> anyhow::Result<()>;
}
```

Components such as the rule enforcer or RPC components would need to implement their migrations manager as `NoOp` since they can decide how they behave internally based on they query the fork manager's currently active spec.

**Comment by rakanalh** [comment]
# Regarding ledger db tables

Since the underlying DB engine is RocksDB, we are dealing with a K/V store which lets us define schemas by defining the types which we need to store into those tables.

In order for us to support new functionality and maintain backward compatibility, old schemas are kept for the sake of RPC calls (as an example) to maintain support for the older versions of the interface and be able to respond to calls based on the values stored up until a specific block prior to activating a new spec which migrates data into a new table.

Assuming a new spec introduces a new ledger DB schema, the ledger implementation would implement the `ForkMigration` in order to migrate a specific table into a new one by inserting data from the old one to the new one with additional data. 

```rust
pub struct LedgerDB { ... etc }

impl ForkMigration for LedgerDB {
  fn pre_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()> {
    let data = self.get_all_soft_confirmations();
    for soft_confirmation in soft_confirmations {
        // get metadata
        let metadata = self.additional_queries_about_soft_confirmation(soft_confirmation.l2_range);
        // The old table did not have soft confirmations metadata.
        self.put_data_into_new_soft_confirmations_table(soft_confirmation, metadata)
    } 
  }

  fn post_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()> {
    // Do nothing
  }
}
```

**Comment by eyusufatik** [comment]
>  fn pre_spec_activation(&self, spec_id: SpecId) -> anyhow::Result<()>;

I can't think of anything that can be done here, can you explain?

in the ledger db example as well, can't those things be done in the post function as well? (assuming post function is the first thing called when we reach the height a fork is activated

**Comment by eyusufatik** [comment]
also, will the fork manager be a part of the STF/runtime or something else?

where are these pre/post functions are invoked exactly is what I'm trying to ask

**Comment by rakanalh** [comment]
To be honest, i am on the opposite side where i think the post method would not have any uses cases.

The pre method would be called prior because the tables / data have to be ready before components become operational with the new `SpecId`

**Comment by rakanalh** [comment]
> also, will the fork manager be a part of the STF/runtime or something else?
> 
> where are these pre/post functions are invoked exactly is what I'm trying to ask

I imagine that the fork manager will be part of the sequencer / fullnode as an `Arc<ForkManager>` to be accessible by internal components.

**Comment by eyusufatik** [comment]
> I imagine that the fork manager will be part of the sequencer / fullnode as an Arc<ForkManager> to be accessible by internal components.

Since forks will also affect how things in stf are handled, i think it’s a good idea to have them on the stf level. Or else we won’t be able to use the forks in zk context.

**Comment by rakanalh** [comment]
@eyusufatik 
https://github.com/chainwayxyz/citrea/pull/922

This draft PR implements fork manager and passes it down to sequencer. The same would happen for fullnode / prover.

In this case, given these components' access to the fork manager, they will be able to notify about blocks and query active spec.
For LedgerDB, it's registered as a migration handler to the fork manager to add / remove / migrate data one block prior or one block after the new spec is activated. As for the EVM, we instantiate it on the fly, which means that we can pass down the `SpecId` to the EVM calls in order for the EVM to know how to behave. 

Let me know your thoughts.

**Comment by eyusufatik** [comment]
Closed by #922
